rust/connlib/tunnel/src/p2p_control.rs

```diff
@@ -0,0 +1,149 @@
+//! Firezone's P2P control protocol between clients and gateways.
+
+#[cfg_attr(not(test), expect(dead_code, reason = "Will be used soon."))]
+pub mod setup_dns_resource_nat {
+    use anyhow::{Context, Result};
+    use connlib_model::{DomainName, ResourceId};
+    use ip_packet::{FzP2pControlSlice, IpPacket};
+    use std::net::IpAddr;
+
+    pub const REQ_CODE: u8 = 0;
+    pub const RES_CODE: u8 = 1;
+
+    pub fn request(resource: ResourceId, domain: DomainName, proxy_ips: Vec<IpAddr>) -> IpPacket {
+        debug_assert_eq!(proxy_ips.len(), 8);
+
+        let payload = serde_json::to_vec(&Request {
+            resource,
+            domain,
+            proxy_ips,
+        })
+        .unwrap();
+
+        ip_packet::make::fz_p2p_control([REQ_CODE, 0, 0, 0, 0, 0, 0, 0], &payload)
```

Comment:
What do you think about using a version number byte and then for this version we can use a single byte for the code.

I think it'll be easier to maintain a byte for version.

If we ever need more than a single byte(I don't think this will happen in the near future) we can use the most significant bit to indicate that the next byte is part of the version.

Or we can just use a u32.

Reply:
Hmm, the current byte is effectively a version / type already. It is an indicator on how to interpret the rest of the payload after the header. What advantage do you see in having multiple layers of types? Our header is a fixed 8 bytes in either case.

Reply:
I think that it's easier to check for version compatibility, instead of having to check if the newer version is compatible with all the messages of a previous version we just have to see if it still handles all the messages of the previous version or it doesn't.

Reply:
I don't understand what incompatibility we are testing for with the version? The header is always 8 bytes, I don't think we should make that dynamic based on some version bit.

And all we currently do with the first byte is indicating, how the rest is meant to be parsed (i.e. what the other header bytes mean and what the payload looks like).

Which part of this contract would you like to couple to a version? We can easily create more layers later. For example, if the first byte == 2, we can say the the next byte is another switch on what the payload etc means.

Reply:
Semantically, you can think of the current design as already having two versions: One of the request and one for the response. We have space for another 253 versions. If we think that is enough, then I'd like to leave it at that. If we think that is not enough, I am happy to expand it to 2 bytes, then we have space for 65536 versions.

Reply:
I think I was just mistaken about the semantics of this byte, I was thinking this would change depending on the message if we wanted to add new ones. 

But I think I understand now that this is just whether this message is a request/response

Reply:
You did understand that correctly! The first byte indicates the message type. There is no such thing as a "request / response"-flag.

For each new message, we'll have to change the first byte. New messages define what the rest of the header means and how the rest of the payload is to be interpreted.